Feature/sidebar component

stevenbister opened the pull request:
<!---
Thanks for creating a Pull Request 💖!
-->

### 🧹 Please check off the following before submitting:

- [x] Close issues
- [x] Request reviewers
- [x] Assign yourself
- [ ] Add labels (e.g. `build:osc-ecommerce`)
- [x] Add project
- [x] Set project status
- [x] Add milestone

---

## ☑ List all github issues that this PR will close:

- Closes #1440 
- Closes #1441 
- Closes #1442 

## 📝 Description

Adds the sidebar component for the academic hub

## ⛳️ Current behaviour (updates)

N/a

## 🚀 New behaviour

Adds sidebar for academic hub

## 💣 Is this a breaking change (Yes/No): No

<!-- If Yes, please describe the impact and migration path for users. -->
> Please describe any breaking changes.


## 🧾 Final Checklist
<!-- Only mark what is applicable, Provide additional context if needed in the "Additonal Information". -->

- [ ] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible cross-device impacts (iOS, Android)
- [x] I've considered possible cross-browser impacts (Chrome, Safari, Edge, Firefox)
- [x] I've added tests to cover my changes (Unit, E2E, Integration, Contract, Functional, Acceptance etc)
- [ ] I've added or updated any documentation.


## 📝 Additional Information

> Any additional contextual information, such as relevant links/articles/documentation

<!--
 THANK YOU for your pull request!
-->

stevenbister commented:
> Noticed that this seems to be across the board actually in all tests so don't know if this is something that's crept in somewhere along the line. Perhaps needs a separate ticket to fix at some point. Probably isn't too high on the priority list atm 😅

Ah yeah forgot to mention in the PR that i'd opened a PR to fix all of this https://github.com/Open-Study-College/osc/pull/1469 will rebase this one and that should get them sorted

stevenbister commented:
Thanks man, all rebased so tests should be working again 🤞 

> Couple of comments:
> 
> * There's a few accessibility violations in Storybook. In desktop mode you're getting the following issue on the Sub Item story:
>   `IDs used in ARIA and labels must be unique`

Ah yeah that's an issue with the underlying `Navbar`, accidentally duplicating in the IDs in the close triggers. I've removed the IDs as they don't need to be there anyway so that should fix that one.

>   In mobile view there's a few more across the stories:
>   `Buttons must have discernible text` and `Links must have discernible text`

Lol that was me being lazy in my Stories 😅 I've wrapped the icons with the `AccessibleIcon` component now so should sort that one too

stevenbister commented:
Oh also since I rebased running storybook causes a `useHref` error on this component now -_- 

Will be fixed by #1483 but until then if you just add `isExternal` to the `NavLink` in `SidebarLink` (line 336) for the time being that will let you access it

mattwendzina reviewed (changes requested):
Nice mate. All looking good, just noticed there is a bit of a weird issue with the Sidebar when viewing through the docs though. See screen-recording:
  
https://github.com/Open-Study-College/osc/assets/36510083/fc29c031-4d7e-4df4-97bc-df6b9d172b71

stevenbister commented:
> Nice mate. All looking good, just noticed there is a bit of a weird issue with the Sidebar when viewing through the docs though. See screen-recording:
> 
>  Screen.Recording.2023-08-30.at.11.30.23.mov

Aha whoops! Forgot that `getBoundingClientRect()` takes the position from the viewport not the elements parent 😅 I've updated that to get the position from parent and take it away from the item's position. This needs to sit in each of the functions as it needs to get recalculated each time it's fired.

I've also added a unique ID to the sidebar to use in the selector on line 155, just to handle any occasions where there's more than one on a page at a time - unlikely in an actual app but was contributing to the issues here.

mattwendzina commented:
> > Nice mate. All looking good, just noticed there is a bit of a weird issue with the Sidebar when viewing through the docs though. See screen-recording:
> > Screen.Recording.2023-08-30.at.11.30.23.mov
> 
> Aha whoops! Forgot that `getBoundingClientRect()` takes the position from the viewport not the elements parent 😅 I've updated that to get the position from parent and take it away from the item's position. This needs to sit in each of the functions as it needs to get recalculated each time it's fired.
> 
> I've also added a unique ID to the sidebar to use in the selector on line 155, just to handle any occasions where there's more than one on a page at a time - unlikely in an actual app but was contributing to the issues here.

Nice mate, all looking good now 👍

mattwendzina reviewed (approved):
LGTM
